src/textual/_time_to_completion.py

```diff
@@ -0,0 +1,301 @@
+"""Code to help calculate the estimated time to completion of some process."""
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+from time import monotonic
+
+from rich.repr import Result
+from typing_extensions import Self
+
+
+@dataclass
+class Sample:
+    """A sample."""
+
+    value: float
+    """The value of the sample."""
+
+    moment: float
+    """The moment when the sample was taken."""
+
+
+class Samples:
+    """A deque-ish-like object that holds samples.
+
+    The window for the samples is based on time, with a maximum sample cap
+    within that window to ensure memory use doesn't run away. Also, for
+    times when the sample window starts to run dry (drops to just the one
+    available sample in the window) the [last-dropped
+    sample][textual._time_to_completion.Samples.last_dropped] is held on to
+    and made available.
+    """
+
+    def __init__(self, time_window_size: float, max_samples: int) -> None:
+        """Initialise the samples object.
+
+        Args:
+            time_window_size: The window of time to keep samples for.
+            max_samples: The maximum number of samples to keep.
+        """
+        self._time_window_size = time_window_size
+        """The maximum amount of time to keep the samples for."""
+        self._max_samples = max_samples
+        """The maximum number of samples to keep."""
+        self._samples: list[Sample] = []
+        """The samples."""
+        self._last_dropped: Sample | None = None
+        """Keep track of the last-dropped sample."""
+
+    def _prune(self) -> Self:
+        """Prune the samples.
+
+        Returns:
+            Self.
+        """
+        if samples := self._samples:
+            # Trim off any "too old" samples.
+            oldest_time = samples[-1].moment - self._time_window_size
```

Comment:
Should the pruning really occur from the last sample, or should it occur from the current time?

Reply:
It makes sense to me that it works off the last-known sample rather than off "now", as what we're making here is a heuristic for figuring out a likely time to completion, and going off "last best" makes more sense to me than going off nothing.